OpenRA.Mods.Common/Widgets/ProductionPaletteWidget.cs

```diff
@@ -182,60 +182,91 @@ public override bool HandleMouseInput(MouseInput mi)
-				}
-				else if (first != null && first.Paused)
-				{
-					// Resume a paused item
-					Sound.Play(TabClick);
-					World.IssueOrder(Order.PauseProduction(CurrentQueue.Actor, icon.Name, false));
-				}
-				else if (CurrentQueue.BuildableItems().Any(a => a.Name == icon.Name))
-				{
-					// Queue a new item
-					Sound.Play(TabClick);
-					Sound.PlayNotification(World.Map.Rules, World.LocalPlayer, "Speech", CurrentQueue.Info.QueuedAudio, World.LocalPlayer.Country.Race);
-					World.IssueOrder(Order.StartProduction(CurrentQueue.Actor, icon.Name,
-						handleMultiple ? 5 : 1));
-				}
-				else
-					Sound.Play(DisabledTabClick);
+				Sound.Play(TabClick);
+				World.OrderGenerator = new PlaceBuildingOrderGenerator(CurrentQueue, icon.Name);
+				return true;
 			}
-			else
+
+			return false;
+		}
+
+		public bool PickUpCompletedBuilding()
+		{
+			foreach (var icon in icons.Values)
 			{
-				// Hold/Cancel an existing item
-				if (first != null)
-				{
-					Sound.Play(TabClick);
+				var item = icon.Queued.FirstOrDefault();
+				if (PickUpCompletedBuildingIcon(icon, item))
+					return true;
+			}
 
-					// instant cancel of things we havent started yet and things that are finished
-					if (first.Paused || first.Done || first.TotalCost == first.RemainingCost)
-					{
-						Sound.PlayNotification(World.Map.Rules, World.LocalPlayer, "Speech", CurrentQueue.Info.CancelledAudio, World.LocalPlayer.Country.Race);
-						World.IssueOrder(Order.CancelProduction(CurrentQueue.Actor, icon.Name,
-							handleMultiple ? 5 : 1));
-					}
-					else
-					{
-						Sound.PlayNotification(World.Map.Rules, World.LocalPlayer, "Speech", CurrentQueue.Info.OnHoldAudio, World.LocalPlayer.Country.Race);
-						World.IssueOrder(Order.PauseProduction(CurrentQueue.Actor, icon.Name, true));
-					}
-				}
-				else
-					Sound.Play(DisabledTabClick);
+			return false;
+		}
+
+		bool HandleLeftClick(ProductionItem item, ProductionIcon icon, bool handleMultiple)
+		{
+			if (PickUpCompletedBuildingIcon(icon, item))
+				return true;
+
+			if (item != null && item.Paused)
+			{
+				// Resume a paused item
+				Sound.Play(TabClick);
+				World.IssueOrder(Order.PauseProduction(CurrentQueue.Actor, icon.Name, false));
+				return true;
 			}
 
+			if (CurrentQueue.BuildableItems().Any(a => a.Name == icon.Name))
+			{
+				// Queue a new item
+				Sound.Play(TabClick);
+				Sound.PlayNotification(World.Map.Rules, World.LocalPlayer, "Speech", CurrentQueue.Info.QueuedAudio, World.LocalPlayer.Country.Race);
+				World.IssueOrder(Order.StartProduction(CurrentQueue.Actor, icon.Name,
+					handleMultiple ? 5 : 1));
+				return true;
+			}
+
+			return false;
+		}
+
+		bool HandleRightClick(ProductionItem item, ProductionIcon icon, bool handleMultiple)
+		{
+			if (item == null)
+				return false;
+
+			Sound.Play(TabClick);
+
+			// instant cancel of things we havent started yet and things that are finished
```

Comment:
"Instant", "haven't", and also things that are paused. Preferably they should be ordered in che comment like they are in the code.